graphql_compiler/macros/macro_expansion.py

```diff
@@ -28,21 +29,144 @@ def _merge_non_overlapping_dicts(merge_target, new_data):
     return result
 
 
-def _expand_specific_macro_edge(macro_definition_selection_set, selection_ast, subclass_sets=None):
+def _merge_selection_sets(selection_set_a, selection_set_b):
+    """Merge selection sets, merging directives on name conflict."""
+    if selection_set_a is None:
+        return selection_set_b
+    if selection_set_b is None:
+        return selection_set_a
+
+    # Convert to dict
+    selection_dict_a = get_uniquely_named_objects_by_name(selection_set_a.selections)
+    selection_dict_b = get_uniquely_named_objects_by_name(selection_set_b.selections)
+
+    # Compute intersection by name
+    common_selection_dict = dict()
+    common_fields = set(selection_dict_a.keys()) & set(selection_dict_b.keys())
+    for field_name in common_fields:
+        field_a = selection_dict_a[field_name]
+        field_b = selection_dict_b[field_name]
+        if field_a.selection_set is not None or field_b.selection_set is not None:
+            raise GraphQLCompilationError('TODO')
+
+        merged_field = deepcopy(field_a)
```

Comment:
I don't think this needs a `deepcopy`, a shallow one should be fine if you explicitly assign a new `directives` list instead of mutating that one in place. A deep copy is likely to be quite expensive due to the number of nested AST objects that need to be constructed.